svc.go

```diff
@@ -0,0 +1,178 @@
+package autonat
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	pb "github.com/libp2p/go-libp2p-autonat/pb"
+
+	ggio "github.com/gogo/protobuf/io"
+	libp2p "github.com/libp2p/go-libp2p"
+	host "github.com/libp2p/go-libp2p-host"
+	inet "github.com/libp2p/go-libp2p-net"
+	peer "github.com/libp2p/go-libp2p-peer"
+	pstore "github.com/libp2p/go-libp2p-peerstore"
+	ma "github.com/multiformats/go-multiaddr"
+)
+
+const P_CIRCUIT = 290
+
+type AutoNATService struct {
+	ctx    context.Context
+	dialer host.Host
+	peers  map[peer.ID]struct{}
+	mx     sync.Mutex
+}
+
+func NewAutoNATService(ctx context.Context, h host.Host) (*AutoNATService, error) {
+	dialer, err := libp2p.New(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	as := &AutoNATService{
+		ctx:    ctx,
+		dialer: dialer,
+		peers:  make(map[peer.ID]struct{}),
+	}
+	h.SetStreamHandler(AutoNATProto, as.handleStream)
+
+	go as.resetPeers()
+
+	return as, nil
+}
+
+func (as *AutoNATService) handleStream(s inet.Stream) {
+	defer s.Close()
+
+	pid := s.Conn().RemotePeer()
+	log.Debugf("New stream from %s", pid.Pretty())
+
+	r := ggio.NewDelimitedReader(s, inet.MessageSizeMax)
+	w := ggio.NewDelimitedWriter(s)
+
+	var req pb.Message
+	var res pb.Message
+
+	err := r.ReadMsg(&req)
+	if err != nil {
+		s.Reset()
```

Comment:
I feel like we should at least put a debug log here. throwing errors away entirely makes me sad

Reply:
ah yes, you are right; sorry about that.

Reply:
hrm, this can be just EOF from the discovery stream. Do we want to log that?

Reply:
added logging for this https://github.com/libp2p/go-libp2p-autonat/pull/1/commits/b1733eb0cce95ff5a7024cc2f0a8752cd81e0ade

if EOFs end up spamming, we can skip them.